## C:\Users\nitish.mishra\Desktop\eMap\Methylation Data
library(data.table)
library(impute)
library(MethylMix)
x <- fread('TCGA-Methylation-All.csv', sep="auto", sep2="auto", nrows=-1, header = TRUE)
a <- x$V1 #####(This is the list of rownames)
x$V1 <-NULL
rownames(x) <- a ### Now this x is data.frame 
matrix <- data.matrix(x) ### Finally this is the "t" matrix
rm(a, x)
matrixTumor <- matrix[,1:184]
matrixNormal <- matrix[,185:194]
matrixTumor <- matrixTumor[which(rowSums(is.na(matrixTumor)) < 18),]## This will print all those row where =< 19 NA (10% NA of 194)
tmp <- colSums(is.na(matrixTumor))
which(tmp > 39407)### No sample have more than 10% missing NA
matrixNormal <- matrixNormal[which(rowSums(is.na(matrixNormal)) < 2),]
tmp <- colSums(is.na(matrixTumor))
which(tmp > 39560)### No sample have more than 10% missing NA
TumorProbe <- rownames(matrixTumor)
NormalProbe <- rownames(matrixNormal)
commonProbe <- intersect(TumorProbe, NormalProbe)

matrixNormal <- matrixNormal[rownames(matrixNormal) %in% commonProbe,]
matrixTumor <- matrixTumor[rownames(matrixTumor) %in% commonProbe,]
rm(TumorProbe, NormalProbe, commonProbe)

matrixTumor.imputeKNN <- impute.knn(matrixTumor, k = 15, rowmax = 0.10, colmax = 0.10)
matrixTumor.imputeKNN <- matrixTumor.imputeKNN$data
matrixNormal.imputeKNN <- impute.knn(matrixNormal, k = 15, rowmax = 0.10, colmax = 0.10)
matrixNormal.imputeKNN <- matrixNormal.imputeKNN$data
rm(matrixTumor, matrixNormal)

##****************************####################**************************
##Make file Gene-Probe.csv by using gene2probe.pl which have list of gene and its corresponding probes
############################################################################
## gene have twice in Expression file so Delete one with low RSEM values
geneExp <- as.matrix(read.csv(file="PAAD.uncv2.mRNAseq_RSEM_normalized_log2.csv", header=TRUE, sep = ",", row.names = 1, as.is=TRUE))
gene2Probe <- as.matrix(read.csv(file="Gene-Probe-10BPSNP.MAF0.01.csv", header=FALSE, sep = ",", row.names = 1, fill = TRUE))
##"Gene-Probe-New.csv" is CSV file generated by JHU HM450 annotation file by using gene2probe.pl
############################################################################
####---- Remove gene and sample which have more that 10% NA values --#######
geneExp1 <- geneExp[,1:179]
geneExp2 <- geneExp1[which(rowSums(is.na(geneExp1)) < 18),]
tmp <- colSums(is.na(geneExp2))
which(tmp >1846)# TCGA.FB.AAPP.01, geneExp1 <- geneExp1[,-57] 57 i.e. column no 57 have more than 1850  (10% genes) NA so drop thi sample
## But we already removed rows which have more than 10% NA so now column TCGA.FB.AAPP.01 have less than 10% NA
geneExp.imputeKNN <- impute.knn(geneExp2, k = 15, rowmax = 0.10, colmax = 0.10)
geneExp.imputeKNN <- geneExp.imputeKNN$data
rm(geneExp1, geneExp2, tmp)
############################################################################
#Now rum MethylMix for TSS1500 probes, no clustering
# CpGGene2Probe <- gene2Probe[!is.na(gene2Probe)]
# matrix.imputeKNN.probes <- rownames(matrix.imputeKNN)
# common <- intersect(CpGGene2Probe, matrix.imputeKNN.probes)
# impute.tumor.WithGene <- matrixTumor.imputeKNN[rownames(matrixTumor.imputeKNN) %in% common,]
# impute.normal.WithGene <- matrixNormal.imputeKNN[rownames(matrixNormal.imputeKNN) %in% common,]
# matrixTumor.TSS.1.5Kb <- matrixTumor.imputeKNN[rownames(matrixTumor.imputeKNN) %in% TSS1.5KbUPdOwn,]
# matrixNormal.TSS.1.5Kb <- matrixNormal.imputeKNN[rownames(matrixNormal.imputeKNN) %in% TSS1.5KbUPdOwn,]
# MethylResult.TSS.1.5Kb <- MethylMix(matrixTumor.TSS.1.5Kb, matrixNormal.TSS.1.5Kb, geneExp.imputeKNN)
#methgenenames <- unlist(mget(rownames(matrix.imputeKNN), IlluminaHumanMethylation450kSYMBOL, ifnotfound = NA)) ### This command give Symbol for each probe if not then NA
############################################################################
genename <- rownames(gene2Probe)
genenameExp <- rownames(geneExp)
commonGene <- intersect(genename, genenameExp)
gene2Probe.common <- gene2Probe[rownames(gene2Probe) %in% commonGene,]
genename <- rownames(gene2Probe.common)
##################################################################

####Remove all existing files ################
rm(cancer_Clustered, normal_Clustered, cancerClustered, normalClustered, d, m, i, cancer_Clustered1, normal_Clustered1, normal, cancer)
cancer_Clustered=matrix(0,0,length(colnames(matrixTumor.imputeKNN)))
normal_Clustered=matrix(0,0,length(colnames(matrixNormal.imputeKNN)))
for (m in 1:length(genename)) 
{
  d <- gene2Probe.common[m,]
  d[d==""] <- NA
  d <- d[!is.na(d)]
  cancer_Clustered1=matrix(0,0,length(colnames(matrixTumor.imputeKNN)))
  normal_Clustered1=matrix(0,0,length(colnames(matrixNormal.imputeKNN)))
  if(length(d) >1){
  normal <- matrixNormal.imputeKNN[rownames(matrixNormal.imputeKNN) %in% d,]
  cancer <- matrixTumor.imputeKNN[rownames(matrixTumor.imputeKNN) %in% d,]
  ProbeCorrelation=cor(t(cancer),method='pearson')
  ClusterResults=hclust(as.dist(1-ProbeCorrelation), method = "complete", members = NULL)
  cancerClustered=matrix(0,0,length(colnames(cancer)))
  normalClustered=matrix(0,0,length(colnames(normal)))
  Clusternames=c()
  Clusters=cutree(ClusterResults,h=0.75)
  #Clusters=cutree(ClusterResults,h=0.7)
  GeneProbes=rownames(cancer)
  for (i in 1:length(unique(Clusters))) {
    tmpGeneProbes=GeneProbes[Clusters==i]
    #print(tmpGeneProbes)
    if (length(tmpGeneProbes)>1) {
      tmpAveragedProfile=colMeans(cancer[tmpGeneProbes,])
      cancerClustered=rbind(cancerClustered, tmpAveragedProfile)          
      # Same for normal
      #print(length(tmpGeneProbes))
      tmpAveragedProfile=colMeans(normal[tmpGeneProbes,])
      normalClustered=rbind(normalClustered, tmpAveragedProfile)
    } else {          
      cancerClustered=rbind(cancerClustered, cancer[tmpGeneProbes,])
      normalClustered=rbind(normalClustered, normal[tmpGeneProbes,])
      #print(length(tmpGeneProbes))
    }
    Clusternames=c(Clusternames,paste(genename[m],i,sep="-"))
  }
  rownames(cancerClustered)=Clusternames
  rownames(normalClustered)=Clusternames
  normal_Clustered <- rbind(normal_Clustered, normalClustered)
  cancer_Clustered <- rbind(cancer_Clustered, cancerClustered)
  }
  else{
    normal_Clustered1=rbind(normal_Clustered1, matrixNormal.imputeKNN[rownames(matrixNormal.imputeKNN) %in% d,])
    cancer_Clustered1=rbind(cancer_Clustered1, matrixTumor.imputeKNN[rownames(matrixTumor.imputeKNN) %in% d,])
    rownames(normal_Clustered1) <- commonGene[m]
    rownames(cancer_Clustered1) <- commonGene[m]
    normal_Clustered <- rbind(normal_Clustered, normal_Clustered1)
    cancer_Clustered <- rbind(cancer_Clustered, cancer_Clustered1)
  }
}
###########################################################################################################################
#### Trim the colname, so name became similar ##########################
colnames(geneExp.imputeKNN.commonGene) <- gsub("\\.", "-", colnames(geneExp.imputeKNN.commonGene))
colnames(cancer_Clustered) <- substr(colnames(cancer_Clustered),1,15)
colnames(normal_Clustered) <- substr(colnames(normal_Clustered),1,15)
########################################################################
geneExp.imputeKNN.commonGene <- geneExp.imputeKNN[rownames(geneExp.imputeKNN) %in% commonGene,]
MethylMixResults=MethylMix(cancer_Clustered, normal_Clustered, geneExp.imputeKNN.commonGene,OutputRoot= '', Parallel = TRUE)
MethylMix_PlotModel('CDH1---Cluster1' ,METcancer_CDH1_Clustered,MethylMixResults, MAcancer_CDH1,METnormal_CDH1_Clustered)

############################ In PDAC I foud sixteen driver genes ##############
MethylMix_PlotModel('CARD16', cancer_Clustered, MethylMixResults)
MethylMix_PlotModel('DHRS2', cancer_Clustered, MethylMixResults)
MethylMix_PlotModel('FGL2', cancer_Clustered, MethylMixResults)
MethylMix_PlotModel('GBP2', cancer_Clustered, MethylMixResults)
MethylMix_PlotModel('LOC541471', cancer_Clustered, MethylMixResults)
MethylMix_PlotModel('LRRIQ1', cancer_Clustered, MethylMixResults)
MethylMix_PlotModel('PTPLAD2',cancer_Clustered,MethylMixResults,MAdata = geneExp.imputeKNN.commonGene)# Plot of Methylation and gene expression