library(data.table)
library(impute)
library(MethylMix)
x <- fread('TCGA-Methylation-All.csv', sep="auto", sep2="auto", nrows=-1, header = TRUE)
a <- x$V1 #####(This is the list of rownames)
x$V1 <-NULL
rownames(x) <- a ### Now this x is data.frame 
matrix <- data.matrix(x) ### Finally this is the "t" matrix
rm(a, x)
#y <- c(rep(0,194))
#for(i in 1:194){(y[i] <- sum(is.na(matrix[,i])))}
#matrix1 <- matrix[which(rowSums(is.na(matrix))==0),]###Martix where rows have no NA---- This will allow only line which have no NA
##matrix1 <- matrix[which(rowSums(!is.na(matrix))==0),] matrix where all row have only NA
matrixTumor <- matrix[,1:184]
matrixNormal <- matrix[,185:194]
matrixTumor <- matrixTumor[which(rowSums(is.na(matrixTumor)) < 18),]## This will print all those row where =< 19 NA (10% NA of 194)
tmp <- colSums(is.na(matrixTumor))
which(tmp > 39407)### No psample have more than 10% missing NA
matrixNormal <- matrixNormal[which(rowSums(is.na(matrixNormal)) < 2),]
tmp <- colSums(is.na(matrixTumor))
which(tmp > 39560)
TumorProbe <- rownames(matrixTumor)
NormalProbe <- rownames(matrixNormal)
commonProbe <- intersect(TumorProbe, NormalProbe)

matrixNormal <- matrixNormal[rownames(matrixNormal) %in% commonProbe,]
rm(TumorProbe, NormalProbe, commonProbe)
#cross-check with AWK command
# awk -F "NA" '{print $0,NF-1}' Methylation-01A-011A.csv >Methylation-All-1_NA.csv
# awk 'BEGIN{FS=",";}{print $1"\t"$196;}' Methylation-All-1_NA.csv |awk '($2 <= 19)'|wc -l
matrixTumor.imputeKNN <- impute.knn(matrixTumor, k = 15, rowmax = 0.10, colmax = 0.10)
matrixTumor.imputeKNN <- matrixTumor.imputeKNN$data
matrixNormal.imputeKNN <- impute.knn(matrixNormal, k = 15, rowmax = 0.10, colmax = 0.10)
matrixNormal.imputeKNN <- matrixNormal.imputeKNN$data
rm(matrixTumor, matrixNormal)
#############################
#library(IlluminaHumanMethylation450k.db)
# x <- IlluminaHumanMethylation450kSYMBOL
# mapped_probes <- mappedkeys(x)
# xx <- as.list(x[mapped_probes])
# uniqGeneSymbol <- unlist(unique(xx))
# #a <- names(unlist(grep("^PTEN$", xx, value=TRUE)))
# #names(xx[xx %in% c(uniqGeneSymbol[1])])
# #names(unlist(grep("ACTN1", xx, value=TRUE)))##Both line are same
# gene.probe <- matrix(data=NA,nrow=19334,ncol=1500)
# gene.probe <- matrix(data=NA,nrow=19334,ncol=1500)
# rownames(gene.probe) <- uniqGeneSymbol
# for(i in 1:19334)
# {m <- names(xx[xx %in% c(uniqGeneSymbol[i])])
# p<- length(m)
# gene.probe[i,1:p] <- m}
# gene.probe <- gene.probe[,colSums(is.na(gene.probe))<nrow(gene.probe)] 
# rm(x, xx, p, m)
##****************************####################**************************
##I make file Gene-Probe.csv by using gene2probe.pl which have more probes for more genes
############################################################################
##SLC35E2 gene have twice in Expression file so Delete one with low RSEM values
geneExp <- as.matrix(read.csv(file="PAAD.uncv2.mRNAseq_RSEM_normalized_log2.csv", header=TRUE, sep = ",", row.names = 1, as.is=TRUE))
gene2Probe <- as.matrix(read.csv(file="Gene-Probe.csv", header=FALSE, sep = ",", row.names = 1, fill = TRUE))
##"Gene-Probe.csv" is CSV file generated by JHU HM450 annotation file by using gene2probe.pl
############################################################################
####---- Remove gene and sample which have more that 10% NA values --#######
geneExp1 <- geneExp[,1:179]
geneExp2 <- geneExp1[which(rowSums(is.na(geneExp1)) < 18),]
tmp <- colSums(is.na(geneExp2))
which(tmp >1846)# TCGA.FB.AAPP.01, geneExp1 <- geneExp1[,-57] 57 i.e. column no 57 have more than 1850  (10% genes) NA so drop thi sample
## But we already removed rows which have more than 10% NA so now column TCGA.FB.AAPP.01 have less than 10% NA
geneExp.imputeKNN <- impute.knn(geneExp2, k = 15, rowmax = 0.10, colmax = 0.10)
geneExp.imputeKNN <- geneExp.imputeKNN$data
rm(geneExp1, geneExp2, tmp)
############################################################################
#Now rum MethylMix for TSS1500 probes, no clustering
matrixTumor.TSS.1.5Kb <- matrixTumor.imputeKNN[rownames(matrixTumor.imputeKNN) %in% TSS1.5KbUPdOwn,]
matrixNormal.TSS.1.5Kb <- matrixNormal.imputeKNN[rownames(matrixNormal.imputeKNN) %in% TSS1.5KbUPdOwn,]
MethylResult.TSS.1.5Kb <- MethylMix(matrixTumor.TSS.1.5Kb, matrixNormal.TSS.1.5Kb, geneExp.imputeKNN)
#methgenenames <- unlist(mget(rownames(matrix.imputeKNN), IlluminaHumanMethylation450kSYMBOL, ifnotfound = NA)) ### This command give Symbol for each probe if not then NA
############################################################################
genename <- rownames(gene2Probe)
# d <- gene2Probe[1,]
# d <- d[!is.na(d)]
# assign(paste0(genename[1]), d)
rm(cancer_Clustered, normal_Clustered, cancerClustered, normalClustered, d, m, i)
cancer_Clustered=matrix(0,0,length(colnames(matrixTumor.imputeKNN)))
normal_Clustered=matrix(0,0,length(colnames(matrixNormal.imputeKNN)))
for (m in 1:5) 
{
  d <- gene2Probe[m,]
  d <- d[!is.na(d)]
#assign(paste0(genename[i]), d)
#   normal <- matrixNormal.imputeKNN[rownames(matrixNormal.imputeKNN) %in% A1BG,]
#   cancer <- matrixTumor.imputeKNN[rownames(matrixTumor.imputeKNN) %in% A2BG,]
  normal <- matrixNormal.imputeKNN[rownames(matrixNormal.imputeKNN) %in% d,]
  cancer <- matrixTumor.imputeKNN[rownames(matrixTumor.imputeKNN) %in% d,]
  #Expcancer <- geneExp[rownames(geneExp) %in% genename[i],]
  ProbeCorrelation=cor(t(normal),method='pearson')
  ClusterResults=hclust(as.dist(1-ProbeCorrelation), method = "complete", members = NULL)
  cancerClustered=matrix(0,0,length(colnames(cancer)))
  normalClustered=matrix(0,0,length(colnames(normal)))
  #assign(paste0(genename[1]) , Clusters)
  #############################################################
  Clusternames=c()
  Clusters=cutree(ClusterResults,h=0.7)
  GeneProbes=rownames(cancer)
  for (i in 1:length(unique(Clusters))) {
    tmpGeneProbes=GeneProbes[Clusters==i]
    print(tmpGeneProbes)
    if (length(tmpGeneProbes)>1) {
      tmpAveragedProfile=colMeans(cancer[tmpGeneProbes,])
      cancerClustered=rbind(cancerClustered, tmpAveragedProfile)          
    # Same for normal
      print(length(tmpGeneProbes))
      tmpAveragedProfile=colMeans(normal[tmpGeneProbes,])
      normalClustered=rbind(normalClustered, tmpAveragedProfile)
    } else {          
      cancerClustered=rbind(cancerClustered, cancer[tmpGeneProbes,])
      normalClustered=rbind(normalClustered, normal[tmpGeneProbes,])
      print(length(tmpGeneProbes))
    }
    Clusternames=c(Clusternames,paste(genename[m],i,sep="-"))
  }
  rownames(cancerClustered)=Clusternames
  rownames(normalClustered)=Clusternames
  normal_Clustered <- rbind(normal_Clustered, normalClustered)
  cancer_Clustered <- rbind(cancer_Clustered, cancerClustered)
  #write.table(normalClustered, file="Normal_cluster.csv", row.names=TRUE, col.names = TRUE, sep = ",")
  #write.table(cancerClustered, file="Tumor_cluster.csv", row.names=TRUE, col.names = TRUE, sep = ",")
}
MethylMixResults=MethylMix(METcancer_CDH1_Clustered,METnormal_CDH1_Clustered, MAcancer_CDH1,OutputRoot= '', Parallel=FALSE)
MethylMix_PlotModel('CDH1---Cluster1' ,METcancer_CDH1_Clustered,MethylMixResults, MAcancer_CDH1,METnormal_CDH1_Clustered)